Userland/Libraries/LibWeb/HTML/Canvas/CanvasPath.cpp

```diff
@@ -213,4 +213,154 @@ void CanvasPath::rect(double x, double y, double w, double h)
+
+    // 11. Corner curves must not overlap. Scale all radii to prevent this:
+    // 11.1. Let top be upperLeft["x"] + upperRight["x"].
+    // 11.2. Let right be upperRight["y"] + lowerRight["y"].
+    // 11.3. Let bottom be lowerRight["x"] + lowerLeft["x"].
+    // 11.4. Let left be upperLeft["y"] + lowerLeft["y"].
+    // 11.5. Let scale be the minimum value of the ratios w / top, h / right, w / bottom, h / left.
+    double scale = AK::min(AK::min(w / (upper_left.x + upper_right.x), h / (upper_right.y + lower_right.y)), AK::min(w / (lower_right.x + lower_left.x), h / (upper_left.y + lower_left.y)));
+
+    // 11.6. If scale is less than 1, then set the x and y members of upperLeft, upperRight, lowerLeft, and lowerRight to their current values multiplied by scale.
+    if (scale < 1) {
+        upper_left.x *= scale;
+        upper_left.y *= scale;
+        upper_right.x *= scale;
+        upper_right.y *= scale;
+        lower_left.x *= scale;
+        lower_left.y *= scale;
+        lower_right.x *= scale;
+        lower_right.y *= scale;
+    }
+
+    // 12. Create a new subpath:
+    auto transform = active_transform();
+    bool large_arc = false;
+    bool sweep = true;
+
+    // 12.1. Move to the point (x + upperLeft["x"], y).
+    m_path.move_to(transform.map(Gfx::FloatPoint { x + upper_left.x, y }));
+
+    // 12.2. Draw a straight line to the point (x + w − upperRight["x"], y).
+    m_path.line_to(transform.map(Gfx::FloatPoint { x + w - upper_right.x, y }));
+
+    // 12.3. Draw an arc to the point (x + w, y + upperRight["y"]).
+    m_path.elliptical_arc_to(transform.map(Gfx::FloatPoint { x + w, y + upper_right.y }), { upper_right.x, upper_right.y }, transform.rotation(), large_arc, sweep);
+
+    // 12.4. Draw a straight line to the point (x + w, y + h − lowerRight["y"]).
+    m_path.line_to(transform.map(Gfx::FloatPoint { x + w, y + h - lower_right.y }));
+
+    // 12.5. Draw an arc to the point (x + w − lowerRight["x"], y + h).
+    m_path.elliptical_arc_to(transform.map(Gfx::FloatPoint { x + w - lower_right.x, y + h }), { lower_right.x, lower_right.y }, transform.rotation(), large_arc, sweep);
+
+    // 12.6. Draw a straight line to the point (x + lowerLeft["x"], y + h).
+    m_path.line_to(transform.map(Gfx::FloatPoint { x + lower_left.x, y + h }));
+
+    // 12.7. Draw an arc to the point (x, y + h − lowerLeft["y"]).
+    m_path.elliptical_arc_to(transform.map(Gfx::FloatPoint { x, y + h - lower_left.y }), { lower_left.x, lower_left.y }, transform.rotation(), large_arc, sweep);
+
+    // 12.8. Draw a straight line to the point (x, y + upperLeft["y"]).
+    m_path.line_to(transform.map(Gfx::FloatPoint { x, y + upper_left.y }));
+
+    // 12.9. Draw an arc to the point (x + upperLeft["x"], y).
+    m_path.elliptical_arc_to(transform.map(Gfx::FloatPoint { x + upper_left.x, y }), { upper_left.x, upper_left.y }, transform.rotation(), large_arc, sweep);
+
+    // 13. Mark the subpath as closed.
+    m_path.close();
```

Comment:
This seems to close the whole path, not the subpath. Though... I'm not sure we actually support subpaths at all? There's code referring to them in `Gfx::Path` but it's confusing me.

Reply:
Yes I also was confused and don't know if we support it, the current code closes the path and move to the correct closing point so that is following the spec.